main.nf

```diff
@@ -167,22 +245,333 @@ process fastqc {
+        """
+        fastp -w "${task.cpus}" -q "${qual}" --cut_by_quality5 \
+            --cut_by_quality3 --cut_mean_quality "${trim_qual}"\
+            --adapter_sequence=${adapter} --adapter_sequence_r2=${adapter_reverse} \
+            -i "${reads[0]}" -I "${reads[1]}" \
+            -o ${name}_trimmed_R1.fastq.gz -O ${name}_trimmed_R2.fastq.gz
+        """
+    }
+    else {
+        """
+        fastp -w "${task.cpus}" -q "${qual}" --cut_by_quality5 \
+            --cut_by_quality3 --cut_mean_quality "${trim_qual}"\
+            --adapter_sequence="${adapter}" --adapter_sequence_r2="${adapter_reverse}" \
+            -i ${reads} -o "${name}_trimmed.fastq.gz"
+        """
+    }
+}
+
+
+process fastqc_trimmed {
+    tag "$name"
+    publishDir "${params.outdir}/fastqc", mode: 'copy'
+
+    input:
+    set val(name), file(reads) from trimmed_reads_fastqc
+
+    output:
+    file "*_fastqc.{zip,html}" into fastqc_results_trimmed
+
+    script:
+    """
+    fastqc -t "${task.cpus}" -q ${reads}
+    """
+}
+
+
+/*
+ * STEP 2 - Assembly
+ */
+process megahit {
+    tag "$name"
+    publishDir "${params.outdir}/", mode: 'copy'
+
+    input:
+    set val(name), file(reads) from trimmed_reads_megahit
+
+    output:
+    set val(name), file("megahit/${name}.contigs.fa") into (assembly_quast, assembly_metabat, assembly_refinem)
+
+    script:
+    if ( !params.singleEnd ) {
+    """
+    megahit -t "${task.cpus}" -1 "${reads[0]}" -2 "${reads[1]}" -o megahit \
+        --out-prefix "${name}"
+    """
+    }
+    else {
+    """
+    megahit -t "${task.cpus}" -r ${reads} -o megahit --out-prefix "${name}"
+    """
+    }
+}
+
+
+process quast {
+    tag "$name"
+    // publishDir "${params.outdir}/quast/$name", mode: 'copy'
+
+    input:
+    set val(name), file(assembly) from assembly_quast
+
+    output:
+    file("quast_${name}/*") into quast_results
+
+    script:
+    """
+    quast.py --threads "${task.cpus}" -l "${name}" "${assembly}" -o "quast_${name}"
+    """
+}
+
+
+/*
+ * STEP 3 - Binning
+ */
+process metabat {
+    tag "$name"
+    publishDir "${params.outdir}/metabat", mode: 'copy'
+
+    input:
+    set val(_name), file(assembly) from assembly_metabat
+    set val(name), file(reads) from trimmed_reads_metabat
+    val(min_size) from params.min_contig_size
+
+    output:
+    set val(name), file("bins/") into metabat_bins
+    set val(name), file("${name}.bam") into (mapped_reads_checkm, mapped_reads_refinem)
+
+    script:
+    if ( !params.singleEnd ) {
+    """
+    bowtie2-build --threads "${task.cpus}" "${assembly}" ref
+    bowtie2 -p "${task.cpus}" -x ref -1 "${reads[0]}" -2 "${reads[1]}" | \
+        samtools view -@ "${task.cpus}" -bS | \
+        samtools sort -@ "${task.cpus}" -o "${name}.bam"
+    samtools index "${name}.bam"
+    jgi_summarize_bam_contig_depths --outputDepth depth.txt "${name}.bam"
+    metabat2 -t "${task.cpus}" -i "${assembly}" -a depth.txt -o "bins/${name}" -m ${min_size}
+    """
```

Comment:
I'm not sure this does what is intended. I would hope that this rule is executed once for each assembly with all samples mapping against it. Instead, when running with ten samples I get only one mapping sample and it's not necessarily the same as the one assembled, for example:
``` bash
#!/bin/bash -euo pipefail
bowtie2-build --threads "20" "P1994_105.contigs.fa" ref
bowtie2 -p "20" -x ref -1 "P1994_102_trimmed_R1.fastq.gz" -2 "P1994_102_trimmed_R2.fastq.gz" |         samtools view -@ "20" -bS |         samtools sort -@ "20" -o "P1994_102.bam"
samtools index "P1994_102.bam"
jgi_summarize_bam_contig_depths --outputDepth depth.txt "P1994_102.bam"
metabat2 -t "20" -i "P1994_105.contigs.fa" -a depth.txt -o "bins/P1994_102" -m 1500
```